2021-offseason/src/main/java/frc/robot/subsystems/Hood.java

```diff
@@ -0,0 +1,91 @@
+import com.team254.lib.util.InterpolatingTreeMap;
+import com.revrobotics.CANEncoder;
+import com.revrobotics.CANPIDController;
+import com.revrobotics.ControlType;
+
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.controller.PIDController;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.PIDSubsystem;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import org.photonvision.PhotonCamera;
+
+public class Hood extends SubsystemBase {
+    private final CANSparkMax hoodMain;
+    private final CANEncoder hoodEncoder;
+    private final CANPIDController hoodController;
+    private InterpolatingTreeMap closeTreeMap;
+    private InterpolatingTreeMap farTreeMap;
+    private PhotonCamera camera;
+    private double currentDegree = 0;
+    private final DigitalInput limitswitch = new DigitalInput(Constants.kHood.LIMIT_PORT);
+    private boolean isTaring = false;
+    private double initialSetpoint;
+    
+    public Hood() {
+        SmartDashboard.putNumber("kP", Constants.kHood.kP);
+        SmartDashboard.putNumber("kD", Constants.kHood.kD);
+        this.hoodMain = new CANSparkMax(Constants.kHood.MOTOR_ID, Constants.kHood.MOTOR_TYPE);
+        hoodMain.setIdleMode(CANSparkMax.IdleMode.kCoast);
+        this.hoodEncoder = this.hoodMain.getEncoder();
+        this.hoodController = this.hoodMain.getPIDController();
+        this.hoodController.setOutputRange(-Constants.kHood.MAX_SPEED, Constants.kHood.MAX_SPEED);
+        this.hoodController.setP(Constants.kHood.kP);
+        this.hoodController.setI(Constants.kHood.kI);
+        this.hoodController.setD(Constants.kHood.kD);
+        this.camera = new PhotonCamera("myCamera");
+        initialSetpoint = hoodEncoder.getPosition();
+        setZero();
+    }
+    @Override
+    public void periodic() {
+        SmartDashboard.putNumber("hood applied output", hoodMain.getAppliedOutput());
+        SmartDashboard.putNumber("hood encoder position", hoodEncoder.getPosition());
+        if (isTaring) tareHoodPeriodic();
+    }
+
+    public void setZero() {
+        isTaring = true;
+    }
+
+    private void tareHoodPeriodic() {
+        if (limitswitch.get()) {
+            hoodMain.set(0);
+            currentDegree = 0;
+            isTaring = false;
+            initialSetpoint = hoodEncoder.getPosition();
+        }
+        else hoodMain.set(Constants.kHood.MAX_SPEED);
+    }
+
+    public void set(double setpoint) {
+        this.hoodController.setReference(setpoint, ControlType.kPosition);
+    }
+    public void incrementUp() {
+        increaseSetpoint(Constants.kHood.HOOD_INCREMENT);
+    }
+    public void incrementDown() {
+        increaseSetpoint(-Constants.kHood.HOOD_INCREMENT);
+    }
+    public void increaseSetpoint(double amount) {
+        currentDegree+=amount;
+        currentDegree=Math.max(0, currentDegree);
+        set(currentDegree+initialSetpoint);
```

Comment:
nit: spacing between operators
```suggestion
        currentDegree += amount;
        currentDegree = Math.max(0, currentDegree);
        set(currentDegree + initialSetpoint);
```

Reply:
also line 89